framework/db/ExpressionInterface.php

```diff
@@ -0,0 +1,27 @@
+<?php
+
+namespace yii\db;
+
+/**
+ * Interface ExpressionInterface provides API for a DB expression
+ * that does not need additional escaping or quoting.
+ *
+ * The database abstraction layer of Yii framework supports objects that implement this
+ * interface and will use them as for passing raw SQL expressions.
+ *
+ * The default implementation is a class [[Expression]].
+ *
+ * @author Dmytro Naumenko <d.naumenko.a@gmail.com>
+ * @since 2.0.13
+ */
+interface ExpressionInterface
+{
+    /**
+     * Method builds the raw SQL expression
+     *
+     * @param QueryBuilder $queryBuilder the QueryBuilder that intends to use this expression
+     * @param array $params the binding parameters
+     * @return string the raw SQL that will not be additionally escaped or quoted
+     */
+    public function buildUsing(QueryBuilder $queryBuilder, &$params = []);
```

Comment:
I would call it `build()` since QB here is a part of the implementation and `using` in `yii\db` usually means it's somehow related to an `using` keyword/expression.

Reply:
When you see `$expression->build($queryBuilder)`, what is the first meaning you guess?
For me it is something inconsistent like "Build query builder in expression?? Wtf?"
And compare it with `$expression->buildUsing($queryBuilder)`. For me that `Using` improves understanding a lot.

Reply:
Fair enough.

Reply:
Still not happy with the naming, it shoulds like we are building a "using" - whatever that should mean. How about `buildSql()`?

Reply:
First of all I would not pass `QueryBuilder` to this method. In order to be consistent with other DB-relarted code parts, it should accept `yii\db\Connection` instance:

```php
public function build(Connection $db) {...}
```

But the whole interface is inconsistent. If you wish to exrtact an interface for `yii\db\Expression` it should have 2 methods: `getSql()` and `getParams()` instead being built by `QueryBuilder`.
Also DB Expression should have `__toString()` method, allowing it to be typecasted into string for debug purposes. So interface should look like:

```php
interface ExpressionInterface
{
    public function getSql() : string;

    public function getParams() : array;

    public function __toString() : string;
}
```

If expression should be bound to particular DB connection or query builder (while I am unsure why) it shoud be done via constructor parameter or setter.

Reply:
> In order to be consistent with other DB-relarted code parts, it should accept `yii\db\Connection` instance

Expression is related to Query, that is about to build. Not to the connection itself.

> If expression should be bound to particular DB connection or query builder (while I am unsure why) it shoud be done via constructor parameter or setter.

It is not bound, but it can be built within `QueryBuilder`, that belongs to a particular DB connection.
You do not specify neither DB nor `QueryBuilder` when you create a Query, why should we do it for expression. Providing DB connection or `QueryBuilder` via constructor will make the API horrible.

> If you wish to exrtact an interface for yii\db\Expression it should have 2 methods: `getSql()` and `getParams()` instead being built by `QueryBuilder`

Providing **static** SQL and params is only one of the use cases. More general one is providing **some** SQL and params, that shall be used in the concrete query.

> Also DB Expression should have `__toString()` method, allowing it to be typecasted into string for debug purposes. So interface should look like:

I can agree on necessity of Expression dumping for debug purposes, but I am not sure `__toString()` is the best option. Will think more about it.

Reply:
> Still not happy with the naming, it shoulds like we are building a "using" - whatever that should mean. How about `buildSql()`?

It was important for me to indicate causal relationship in this naming: 

I want to build this expression &rarr; I need QueryBuilder and params to do it &rarr; I am building the Expression using QueryBuilder and params &rarr; `$expression->buildUsing($queryBuiler, $params)`

Could you compose something similar with `buildSql()`?

Reply:
> Expression is related to Query, that is about to build. Not to the connection itself.

`Query` accepts `Connection` instance as an argument for final execution methods, e.g. methods `one()`, `all()` and so on accepts `$db` as a connection instance. `QueryBuilder` is just a part of the `Connection`, and actually it is unnecessary part. Particular developer may re-implement Db layer in the way `QueryBuilder` will not be in use.
`QueryBuilder` can always be determined from `Connection` thus using `Connection` is more flexible here.

> You do not specify neither DB nor QueryBuilder when you create a Query, why should we do it for expression. Providing DB connection or QueryBuilder via constructor will make the API horrible.

Expression itself should not depend on Connection or be somehow bound ot it anywhere.
At the present `yii\db\Expression` is used to specify some raw SQL, which should bypass the QueryBuilder. If there is some construction, which should be build depending on particular DB schema - it should be incapsulated inside `QueryBuilder` - not in `Expression`.

> Providing static SQL and params is only one of the use cases. More general one is providing some SQL and params, that shall be used in the concrete query.

I do not understand this part. SQL Expression consists of 2 parts: raw SQL text and binding parameters - nothing else. Thus `ExpressionInterface` should return its raw SQL and binding params related to it, so it should have 2 methods.
It is a taks of `QueryBuilder` to place given SQL text on its place in the one of entire Query and merge binding params. It was always doing this before this PR. Why it should be changed?

> I can agree on necessity of Expression dumping for debug purposes, but I am not sure __toString() is the best option. Will think more about it.

PSR standards using `__toString()` in thier interfaces all the time. Since we accepted several PSR for 2.1, we should pick this approach as well.

Reply:
> `Query` accepts `Connection` instance as an argument for final execution methods

Exactly – for a final methods. In case we add `Connection` as a constructor parameter for `Expression`, we shall pass it much earlier, than calling `one()` or `all()`

> `QueryBuilder` can always be determined from `Connection` thus using `Connection` is more flexible here.

Connection can always be determined from a `QueryBuilder` unambiguously. However, schema can provide different `QueryBuilders` depending on some conditions, so having, `Connection` we can not determine `QueryBuilder` with 100% guarantee.

> Particular developer may re-implement Db layer in the way `QueryBuilder` will not be in use.

I don't think we should consider corner-cases like this.

> Expression itself should not depend on Connection or be somehow bound ot it anywhere.
> ...
> If there is some construction, which should be build depending on particular DB schema - it should be incapsulated inside `QueryBuilder` - not in `Expression`.

I was thinking about it, but have not found intersections with other DBMSs to go this way: MySQL does not support arrays, [Oracle has collections](https://docs.oracle.com/cd/B28359_01/appdev.111/b28370/collections.htm#LNPLS00507) that must be typed (e.g. `user_ids_typ(123,456,789)`) thus `Expression` will have different constructor, No-SQL is a completely different story. 

> PSR standards using `__toString()` in thier interfaces all the time. Since we accepted several PSR for 2.1, we should pick this approach as well.

Ok, I'm fine with it, but what should this method provide in case when `Expression` is not static SQL and `Expression` is not bound to `Connection` or `QueryBuilder`?

Reply:
> MySQL does not support arrays

MySQL has `SET` data type - it is something similar to arrays.

Reply:
See #3796